### PR TITLE
Advertise RDT L3 num_closid

### PR DESCRIPTION
**Pull request by PiotrProkop**
Enable the feature discovery of `RDTL3CA_NUM_CLOSID` as part of the cpu-rdt source.

This feature is intended to be used as `extended` resource when this PR is merged https://github.com/kubernetes-sigs/node-feature-discovery/pull/1099

### CODE REVIEW

**Comment by ArangoGutierrez** [comment]
Hold until #1099
/hold

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]
Could you also add the new features here `docs/usage/features.md`

**Comment by PiotrProkop** [comment]
> Could you also add the new features here `docs/usage/features.md`

@ArangoGutierrez done.

**Review by marquiz — COMMENTED** [review]
Thanks @PiotrProkop for the enhancement! We need to document the new feature in the "Available features" table in customization guide (`docs/usage/customization-guide.md`), need also to change the "feature type" from "flag" to "attribute"

File: source/cpu/rdt_amd64.go [hunk 1]
```diff
@@ -75,17 +82,45 @@ func discoverRDT() []string {
...
+		return -1
+	}
+
+	closidsBytes, err := os.ReadFile(filepath.Join(resctrlRootDir, "info", level, "num_closids"))
+	if err != nil {
+		return -1
+	}
+
+	numClosIDs, err := strconv.ParseInt(string(bytes.TrimSpace(closidsBytes)), 10, 64)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply source/cpu/rdt_amd64.go:120]: Yeah, good idea!

File: docs/usage/features.md [hunk 2]
```diff
@@ -152,14 +152,15 @@ See the full list in [github.com/klauspost/cpuid][klauspost-cpuid].
...
-| RDTMBA    | Intel Memory Bandwidth Allocation (MBA) Technology
+| Flag               | Description                                                                                   |
+| -------------------| ----------------------------------------------------------------------------------------------|
+| RDTMON             | Intel RDT Monitoring Technology
+| RDTCMT             | Intel Cache Monitoring (CMT)
+| RDTMBM             | Intel Memory Bandwidth Monitoring (MBM)
+| RDTL3CA            | Intel L3 Cache Allocation Technology
+| RDTl2CA            | Intel L2 Cache Allocation Technology
+| RDTMBA             | Intel Memory Bandwidth Allocation (MBA) Technology
+| RDTL3CA_NUM_CLOSID | The number or available CLOSid(Class of service ID) for Intel L3 Cache Allocation Technology
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply docs/usage/features.md:163]: I agree that `RDTL3CA_NUM_CLOSID` as a label is of little to no use 😄 will change the code to stop advertise it.

File: source/cpu/rdt_amd64.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply source/cpu/rdt_amd64.go:125]: Debug logs added.

File: docs/usage/features.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply docs/usage/features.md]: removed.

**Comment by PiotrProkop** [comment]
> Thanks @PiotrProkop for the enhancement! We need to document the new feature in the "Available features" table in customization guide (`docs/usage/customization-guide.md`), need also to change the "feature type" from "flag" to "attribute"

Thanks for the review! I have updated `customization-guide`.

File: docs/usage/customization-guide.md [hunk 3]
```diff
@@ -601,8 +601,9 @@ The following features are available for matching:
 |                  |              | **`status`** | string | Status of the driver, possible values are 'active' and 'passive'
 |                  |              | **`turbo`**  | bool   | 'true' if turbo frequencies are enabled, otherwise 'false'
 |                  |              | **`scaling`** | string | Active scaling_governor, possible values are 'powersave' or 'performance'.
-| **`cpu.rdt`**    | flag         |          |            | Intel RDT capabilities supported by the system
+| **`cpu.rdt`**    | attribute    |          |            | Intel RDT capabilities supported by the system
 |                  |              | **`<rdt-flag>`** |    | RDT capability is supported, see [RDT flags](#intel-rdt-flags) for details
+|                  |              | **`RDTL3CA_NUM_CLOSID`** | int  | The number or available CLOSid(Class of service ID) for Intel L3 Cache Allocation Technology
```
Comment (marquiz) [inline docs/usage/customization-guide.md:606]: nit: add space, I'd suggest to write CLOSID in all caps

```suggestion
|                  |              | **`RDTL3CA_NUM_CLOSID`** | int  | The number or available CLOSIDs (Class of service ID) for Intel L3 Cache Allocation Technology
```

File: source/cpu/rdt_amd64.go [hunk 4]
```diff
@@ -75,17 +83,50 @@ func discoverRDT() []string {
...
 		}
 	}
 
-	return features
+	return attributes
+}
+
+func getNumClosID(level string) int64 {
+	resctrlRootDir, err := intelrdt.Root()
+	if err != nil {
```
Comment (marquiz) [inline source/cpu/rdt_amd64.go:107]: Should we log this as well, WDYT?

```suggestion
	if err != nil {
		klog.V(4).ErrorS(err, "failed to determine resctrlfs root directory")
```

File: deployment/helm/values.yaml
Comment (marquiz) [inline deployment/helm/values.yaml:1]: Drop this file

**Review by marquiz — COMMENTED** [review]
Thanks @PiotrProkop, almost there :)

File: deployment/helm/values.yaml
Comment (PiotrProkop) [reply deployment/helm/values.yaml:1]: I should create local `.gitignore` :) removing now

File: docs/usage/customization-guide.md (hunk 3 above)
Comment (PiotrProkop) [reply docs/usage/customization-guide.md:606]: I have based this acronym from this document https://docs.kernel.org/x86/resctrl.html, they use `CLOSid` everywhere. But I have no strong opinion about that.

File: source/cpu/rdt_amd64.go (hunk 4 above)
Comment (PiotrProkop) [reply source/cpu/rdt_amd64.go:107]: Yeah, missed one line.

Comment (PiotrProkop) [reply source/cpu/rdt_amd64.go:107]: fixed

File: deployment/helm/values.yaml
Comment (PiotrProkop) [reply deployment/helm/values.yaml:1]: removed

File: docs/usage/customization-guide.md (hunk 3 above)
Comment (PiotrProkop) [reply docs/usage/customization-guide.md]: changed.

Comment (marquiz) [reply docs/usage/customization-guide.md]: >  they use CLOSid everywhere. But I have no strong opinion about that.

OK, not a big deal. Took a look and they actually use CLOSID, CLOSid and closid at least so 🤷

File: docs/usage/customization-guide.md [hunk 5]
```diff
@@ -601,8 +601,9 @@ The following features are available for matching:
 |                  |              | **`status`** | string | Status of the driver, possible values are 'active' and 'passive'
 |                  |              | **`turbo`**  | bool   | 'true' if turbo frequencies are enabled, otherwise 'false'
 |                  |              | **`scaling`** | string | Active scaling_governor, possible values are 'powersave' or 'performance'.
-| **`cpu.rdt`**    | flag         |          |            | Intel RDT capabilities supported by the system
+| **`cpu.rdt`**    | attribute    |          |            | Intel RDT capabilities supported by the system
 |                  |              | **`<rdt-flag>`** |    | RDT capability is supported, see [RDT flags](#intel-rdt-flags) for details
+|                  |              | **`RDTL3CA_NUM_CLOSID`** | int  | The number or available CLOSID(Class of service ID) for Intel L3 Cache Allocation Technology
```
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply docs/usage/customization-guide.md]: fixed

**Review by marquiz — APPROVED** [review]
Thx @PiotrProkop, looks fine for me now 👍
/assign @ArangoGutierrez

**Review by ArangoGutierrez — APPROVED** [review]
/lgtm

**Comment by marquiz** [comment]
@PiotrProkop I think we could unhold this? Unless you want to verify it first with #1099. I'm ok with both

**Comment by PiotrProkop** [comment]
> @PiotrProkop I think we could unhold this? Unless you want to verify it first with #1099. I'm ok with both

Yeah, we can merge this!

**Pull request merged**